application/src/lib.rs

```diff
@@ -0,0 +1,331 @@
+        match self.0.recv().await {
+            Err(broadcast::error::RecvError::Lagged(_)) => {
+                panic!("Multiple shutdown broadcast requests issued")
+            }
+            Err(broadcast::error::RecvError::Closed) => {
+                panic!("Shutdown channel sender closed prematurely")
+            }
+            Ok(()) => (),
+        }
+    }
+}
+
+/// Subsystem handle.
+///
+/// This allows the user to interact with the subsystem from the outside. Currently, it only
+/// supports calling functions on the subsystem.
+#[derive(Clone)]
+pub struct Subsystem<T> {
+    // Send the subsystem stuff to do.
+    action_tx: mpsc::Sender<Action<T>>,
+}
+
+impl<T: Send + 'static> Subsystem<T> {
+    /// Crate a new subsystem handle.
+    fn new(action_tx: mpsc::Sender<Action<T>>) -> Self {
+        Self { action_tx }
+    }
+
+    /// Dispatch a function call to the subsystem
+    pub async fn call_mut<R: Send + 'static>(
+        &self,
+        func: impl FnOnce(&mut T) -> R + Send + 'static,
+    ) -> R {
+        let (rtx, rrx) = oneshot::channel::<R>();
+
+        self.action_tx
+            .send(Box::new(move |subsys| {
+                let result = func(subsys);
+                rtx.send(result).ok().expect("Value return channel closed");
+            }))
+            .await
+            .ok()
+            .expect("Target subsystem down");
+
+        rrx.await.expect("Target subsystem down")
+    }
+
+    /// Dispatch a function call to the subsystem (convenience method if subsystem is not mutated)
+    pub async fn call<R: Send + 'static>(&self, func: impl FnOnce(&T) -> R + Send + 'static) -> R {
+        self.call_mut(|this| func(this)).await
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    // The substringer passive subsystem (for testing)
+    pub struct Substringer {
+        value: String,
+    }
+
+    impl Substringer {
+        pub fn new(value: String) -> Self {
+            Self { value }
+        }
+
+        pub fn append_get(&mut self, other: &str) -> String {
+            self.value += other;
+            self.value.clone()
+        }
+
+        pub fn substr(&self, begin: usize, end: usize) -> String {
+            self.value.get(begin..end).map_or_else(|| String::new(), str::to_string)
+        }
+
+        pub fn size(&self) -> usize {
+            self.value.len()
+        }
+    }
+
+    // The counter passive subsystem (for testing)
+    pub struct Counter {
+        value: u64,
+    }
+
+    impl Counter {
+        pub fn new() -> Self {
+            Self { value: 13 }
+        }
+
+        pub fn get(&self) -> u64 {
+            self.value
+        }
+
+        pub fn add_and_get(&mut self, amount: u64) -> u64 {
+            self.value += amount;
+            self.value
+        }
+    }
+
+    #[tokio::test]
+    async fn test_passive() {
+        let logpath: Option<&std::path::Path> = None;
+        logging::init_logging(logpath);
+
+        let app = Application::new("app");
```

Comment:
Can you please also add a test for non-passive things, where subsystems interact with one another?